File: plan/validator.py
"""Dependency validation for project plans."""

from dataclasses import dataclass, field
from typing import List, Dict, Set, Optional

from .models import ProjectPlan, Task, Story, Epic, TaskStatus


@dataclass
class ValidationError:
    """A validation error in the plan."""
    level: str  # "error" or "warning"
    entity_type: str  # "task", "story", "epic", "milestone"
    entity_id: str
    message: str

    def __str__(self) -> str:
        return f"[{self.level.upper()}] {self.entity_type} {self.entity_id}: {self.message}"


@dataclass
class ValidationResult:
    """Result of plan validation."""
    is_valid: bool = True
    errors: List[ValidationError] = field(default_factory=list)
    warnings: List[ValidationError] = field(default_factory=list)

    def add_error(self, entity_type: str, entity_id: str, message: str):
        """Add an error."""
        self.errors.append(ValidationError("error", entity_type, entity_id, message))
        self.is_valid = False

    def add_warning(self, entity_type: str, entity_id: str, message: str):
        """Add a warning."""
        self.warnings.append(ValidationError("warning", entity_type, entity_id, message))


class PlanValidator:
    """Validates project plan integrity and dependencies."""

    def validate(self, plan: ProjectPlan) -> ValidationResult:
        """Run all validations on a project plan."""
        result = ValidationResult()

        self._validate_task_dependencies(plan, result)
        self._validate_story_dependencies(plan, result)
        self._validate_epic_dependencies(plan, result)
        self._validate_milestone_epics(plan, result)
        self._validate_task_roles(plan, result)
        self._validate_acceptance_criteria(plan, result)
        self._check_circular_dependencies(plan, result)

        return result

    def _validate_task_dependencies(self, plan: ProjectPlan, result: ValidationResult):
        """Validate that all task dependencies exist."""
        all_task_ids = {t.id for t in plan.get_all_tasks()}

        for task in plan.get_all_tasks():
            for dep_id in task.dependencies:
                if dep_id not in all_task_ids:
                    result.add_error(
                        "task", task.id,
                        f"Dependency '{dep_id}' does not exist"
                    )

    def _validate_story_dependencies(self, plan: ProjectPlan, result: ValidationResult):
        """Validate that all story dependencies exist."""
        all_story_ids = {s.id for s in plan.get_all_stories()}

        for story in plan.get_all_stories():
            for dep_id in story.dependencies:
                if dep_id not in all_story_ids:
                    result.add_error(
                        "story", story.id,
                        f"Dependency '{dep_id}' does not exist"
                    )

            for block_id in story.blocks:
                if block_id not in all_story_ids:
                    result.add_warning(
                        "story", story.id,
                        f"Blocks reference '{block_id}' does not exist"
                    )

    def _validate_epic_dependencies(self, plan: ProjectPlan, result: ValidationResult):
        """Validate that all epic dependencies exist."""
        all_epic_ids = {e.id for e in plan.epics}

        for epic in plan.epics:
            for dep_id in epic.dependencies:
                if dep_id not in all_epic_ids:
                    result.add_error(
                        "epic", epic.id,
                        f"Dependency '{dep_id}' does not exist"
                    )

    def _validate_milestone_epics(self, plan: ProjectPlan, result: ValidationResult):
        """Validate that all milestone epic references exist."""
        all_epic_ids = {e.id for e in plan.epics}

        for milestone in plan.milestones:
            for epic_id in milestone.epic_ids:
                if epic_id not in all_epic_ids:
                    result.add_error(
                        "milestone", milestone.id,
                        f"Epic reference '{epic_id}' does not exist"
                    )

    def _validate_task_roles(self, plan: ProjectPlan, result: ValidationResult):
        """Validate that all task roles have corresponding personas."""
        all_roles = {p.role for p in plan.personas}

        for task in plan.get_all_tasks():
            if task.role and task.role not in all_roles:
                result.add_warning(
                    "task", task.id,
                    f"Role '{task.role}' has no corresponding persona"
                )

    def _validate_acceptance_criteria(self, plan: ProjectPlan, result: ValidationResult):
        """Validate that stories have acceptance criteria."""
        for story in plan.get_all_stories():
            if not story.acceptance_criteria:
                result.add_warning(
                    "story", story.id,
                    "Story has no acceptance criteria"
                )

            # Check that test scenarios map to acceptance criteria
            ac_ids = {ac.id for ac in story.acceptance_criteria}
            for ts in story.test_scenarios:
                if ts.acceptance_criterion_id and ts.acceptance_criterion_id not in ac_ids:
                    result.add_warning(
                        "story", story.id,
                        f"Test scenario {ts.id} references non-existent acceptance criterion {ts.acceptance_criterion_id}"
                    )

    def _check_circular_dependencies(self, plan: ProjectPlan, result: ValidationResult):
        """Check for circular dependencies in tasks."""
        # Build dependency graph
        task_graph: Dict[str, Set[str]] = {}
        for task in plan.get_all_tasks():
            task_graph[task.id] = set(task.dependencies)

        # Check for cycles using DFS
        visited: Set[str] = set()
        rec_stack: Set[str] = set()

        def has_cycle(task_id: str, path: List[str]) -> Optional[List[str]]:
            visited.add(task_id)
            rec_stack.add(task_id)

            for dep_id in task_graph.get(task_id, set()):
                if dep_id not in visited:
                    cycle = has_cycle(dep_id, path + [task_id])
                    if cycle:
                        return cycle
                elif dep_id in rec_stack:
                    return path + [task_id, dep_id]

            rec_stack.remove(task_id)
            return None

        for task_id in task_graph:
            if task_id not in visited:
                cycle = has_cycle(task_id, [])
                if cycle:
                    cycle_str = " -> ".join(cycle)
                    result.add_error(
                        "task", task_id,
                        f"Circular dependency detected: {cycle_str}"
                    )

    def get_dependency_order(self, plan: ProjectPlan) -> List[Task]:
        """
        Return tasks in dependency order (topological sort).
        Tasks with no dependencies come first.
        """
        # Build dependency graph
        task_graph: Dict[str, Set[str]] = {}
        all_tasks = {t.id: t for t in plan.get_all_tasks()}

        for task_id, task in all_tasks.items():
            task_graph[task_id] = set(task.dependencies)

        # Topological sort using Kahn's algorithm
        in_degree: Dict[str, int] = {tid: 0 for tid in task_graph}
        for deps in task_graph.values():
            for dep in deps:
                if dep in in_degree:
                    in_degree[dep] += 1

        # Start with tasks that have no incoming edges (nothing depends on them)
        # Actually we want tasks with no dependencies first
        queue = [tid for tid, deps in task_graph.items() if not deps]
        result = []

        while queue:
            task_id = queue.pop(0)
            result.append(all_tasks[task_id])

            # Find tasks that depend on this one
            for tid, deps in task_graph.items():
                if task_id in deps:
                    deps.remove(task_id)
                    if not deps and tid not in result and tid not in queue:
                        queue.append(tid)

        return result

    def get_blocked_tasks(self, plan: ProjectPlan) -> Dict[str, List[str]]:
        """
        Return a dict of task_id -> list of blocking task IDs.
        Only includes tasks that are actually blocked.
        """
        completed_ids = {
            t.id for t in plan.get_all_tasks()
            if t.status == TaskStatus.COMPLETE
        }

        blocked = {}
        for task in plan.get_all_tasks():
            if task.status != TaskStatus.COMPLETE:
                blocking = [d for d in task.dependencies if d not in completed_ids]
                if blocking:
                    blocked[task.id] = blocking

        return blocked

    def get_available_tasks(self, plan: ProjectPlan) -> List[Task]:
        """Get all tasks that are available to work on (no blockers)."""
        blocked = self.get_blocked_tasks(plan)

        available = []
        for task in plan.get_all_tasks():
            if task.status == TaskStatus.AVAILABLE and task.id not in blocked:
                available.append(task)

        return available

    def get_available_tasks_for_role(self, plan: ProjectPlan, role: str) -> List[Task]:
        """Get available tasks for a specific role."""
        return [t for t in self.get_available_tasks(plan) if t.role == role]

File: src/plan/validator.js
/**
 * @file Plan validation logic.
 * Validates project plans for consistency and completeness.
 * @module plan/validator
 */

import { TaskStatus } from '../types/index.js';

/**
 * @typedef {Object} ValidationResult
 * @property {boolean} isValid - Whether validation passed
 * @property {string[]} errors - List of error messages
 * @property {string[]} warnings - List of warning messages
 */

/**
 * Validator for project plans.
 */
export class PlanValidator {
  /**
   * Create a PlanValidator.
   */
  constructor() {
    /** @private @type {string[]} */
    this._errors = [];
    /** @private @type {string[]} */
    this._warnings = [];
  }

  /**
   * Validate a project plan.
   * @param {import('./models.js').ProjectPlan} plan - Plan to validate
   * @returns {ValidationResult}
   */
  validate(plan) {
    this._errors = [];
    this._warnings = [];

    this._validateBasicStructure(plan);
    this._validateTaskDependencies(plan);
    this._validatePersonas(plan);
    this._validateMilestones(plan);
    this._validateUniqueIds(plan);
    this._validateTaskRoles(plan);

    return {
      isValid: this._errors.length === 0,
      errors: [...this._errors],
      warnings: [...this._warnings],
    };
  }

  /**
   * Validate basic plan structure.
   * @private
   * @param {import('./models.js').ProjectPlan} plan
   */
  _validateBasicStructure(plan) {
    if (!plan.name) {
      this._errors.push('Plan must have a name');
    }

    if (plan.epics.length === 0) {
      this._warnings.push('Plan has no epics');
    }

    for (const epic of plan.epics) {
      if (!epic.id) {
        this._errors.push('Epic must have an ID');
      }
      if (!epic.title) {
        this._errors.push(`Epic ${epic.id || '(no ID)'} must have a title`);
      }

      if (epic.stories.length === 0) {
        this._warnings.push(`Epic ${epic.id} has no stories`);
      }

      for (const story of epic.stories) {
        if (!story.id) {
          this._errors.push('Story must have an ID');
        }
        if (!story.title) {
          this._errors.push(`Story ${story.id || '(no ID)'} must have a title`);
        }

        if (story.tasks.length === 0) {
          this._warnings.push(`Story ${story.id} has no tasks`);
        }

        for (const task of story.tasks) {
          if (!task.id) {
            this._errors.push('Task must have an ID');
          }
          if (!task.description) {
            this._errors.push(`Task ${task.id || '(no ID)'} must have a description`);
          }
          if (!task.role) {
            this._errors.push(`Task ${task.id} must have a role`);
          }
        }
      }
    }
  }

  /**
   * Validate task dependencies.
   * @private
   * @param {import('./models.js').ProjectPlan} plan
   */
  _validateTaskDependencies(plan) {
    const allTaskIds = new Set(plan.getAllTasks().map((t) => t.id));

    for (const task of plan.getAllTasks()) {
      for (const depId of task.dependencies) {
        if (!allTaskIds.has(depId)) {
          this._errors.push(`Task ${task.id} has unknown dependency: ${depId}`);
        }

        if (depId === task.id) {
          this._errors.push(`Task ${task.id} depends on itself`);
        }
      }
    }

    // Check for circular dependencies
    const visited = new Set();
    const recursionStack = new Set();

    const hasCycle = (taskId, path = []) => {
      if (recursionStack.has(taskId)) {
        const cycleStart = path.indexOf(taskId);
        const cycle = path.slice(cycleStart).concat(taskId);
        this._errors.push(`Circular dependency detected: ${cycle.join(' -> ')}`);
        return true;
      }

      if (visited.has(taskId)) {
        return false;
      }

      visited.add(taskId);
      recursionStack.add(taskId);

      const task = plan.getTaskById(taskId);
      if (task) {
        for (const depId of task.dependencies) {
          if (hasCycle(depId, [...path, taskId])) {
            return true;
          }
        }
      }

      recursionStack.delete(taskId);
      return false;
    };

    for (const task of plan.getAllTasks()) {
      visited.clear();
      recursionStack.clear();
      hasCycle(task.id);
    }
  }

  /**
   * Validate personas configuration.
   * @private
   * @param {import('./models.js').ProjectPlan} plan
   */
  _validatePersonas(plan) {
    if (plan.personas.length === 0) {
      this._warnings.push('Plan has no personas defined');
      return;
    }

    const roles = new Set();
    for (const persona of plan.personas) {
      if (!persona.id) {
        this._errors.push('Persona must have an ID');
      }
      if (!persona.role) {
        this._errors.push(`Persona ${persona.id || '(no ID)'} must have a role`);
      }
      if (roles.has(persona.role)) {
        this._warnings.push(`Duplicate persona role: ${persona.role}`);
      }
      roles.add(persona.role);
    }
  }

  /**
   * Validate milestones configuration.
   * @private
   * @param {import('./models.js').ProjectPlan} plan
   */
  _validateMilestones(plan) {
    const epicIds = new Set(plan.epics.map((e) => e.id));

    for (const milestone of plan.milestones) {
      if (!milestone.id) {
        this._errors.push('Milestone must have an ID');
      }
      if (!milestone.name) {
        this._errors.push(`Milestone ${milestone.id || '(no ID)'} must have a name`);
      }

      for (const epicId of milestone.epicIds) {
        if (!epicIds.has(epicId)) {
          this._errors.push(`Milestone ${milestone.id} references unknown epic: ${epicId}`);
        }
      }

      if (milestone.epicIds.length === 0) {
        this._warnings.push(`Milestone ${milestone.id} has no epics`);
      }
    }

    // Check for epics not in any milestone
    const milestonedEpics = new Set(plan.milestones.flatMap((m) => m.epicIds));
    for (const epic of plan.epics) {
      if (!milestonedEpics.has(epic.id)) {
        this._warnings.push(`Epic ${epic.id} is not in any milestone`);
      }
    }
  }

  /**
   * Validate unique IDs across all entities.
   * @private
   * @param {import('./models.js').ProjectPlan} plan
   */
  _validateUniqueIds(plan) {
    const ids = new Map();

    const checkId = (type, id) => {
      if (ids.has(id)) {
        this._errors.push(`Duplicate ID: ${id} (${type} and ${ids.get(id)})`);
      } else {
        ids.set(id, type);
      }
    };

    for (const epic of plan.epics) {
      checkId('Epic', epic.id);
      for (const story of epic.stories) {
        checkId('Story', story.id);
        for (const task of story.tasks) {
          checkId('Task', task.id);
        }
        for (const ac of story.acceptanceCriteria) {
          checkId('AcceptanceCriterion', ac.id);
        }
      }
    }

    for (const milestone of plan.milestones) {
      checkId('Milestone', milestone.id);
    }

    for (const persona of plan.personas) {
      checkId('Persona', persona.id);
    }
  }

  /**
   * Validate that task roles match defined personas.
   * @private
   * @param {import('./models.js').ProjectPlan} plan
   */
  _validateTaskRoles(plan) {
    if (plan.personas.length === 0) {
      return; // Can't validate if no personas defined
    }

    const roles = new Set(plan.personas.map((p) => p.role));

    for (const task of plan.getAllTasks()) {
      if (!roles.has(task.role)) {
        this._warnings.push(`Task ${task.id} has role "${task.role}" which is not defined in personas`);
      }
    }
  }

  /**
   * Quick validation check - returns true/false without details.
   * @param {import('./models.js').ProjectPlan} plan - Plan to validate
   * @returns {boolean}
   */
  isValid(plan) {
    return this.validate(plan).isValid;
  }
}
